tsdb/db.go

```diff
@@ -1505,34 +1505,41 @@ func (db *DB) CleanTombstones() (err error) {
 	start := time.Now()
 	defer db.metrics.tombCleanTimer.Observe(time.Since(start).Seconds())
 
-	newUIDs := []ulid.ULID{}
-	defer func() {
-		// If any error is caused, we need to delete all the new directory created.
-		if err != nil {
-			for _, uid := range newUIDs {
+	cleanUpCompleted := false
+	// Repeat cleanup until there is no tombstones left.
+	for !cleanUpCompleted {
+		cleanUpCompleted = true
+
+		for _, pb := range db.Blocks() {
+			uid, safeToDelete, cleanErr := pb.CleanTombstones(db.Dir(), db.compactor)
+			if cleanErr != nil {
+				return errors.Wrapf(cleanErr, "clean tombstones: %s", pb.Dir())
+			}
+			if !safeToDelete {
+				// There was nothing to clean.
+				continue
+			}
+
+			// In case tombstones of the old block covers the whole block,
+			// then there would be no resultant block to tell the parent.
+			pb.meta.Compaction.Deletable = safeToDelete
+			cleanUpCompleted = false
+			if err = db.reload(); err == nil { // Will try to delete old block.
```

Comment:
I think this is not enough to delete the block?

Reply:
Reloading is supposed to delete obsolete blocks, such as the ones that were already cleaned, and add the new block to the db object.

If we fail to delete an obsolete block, for whatever reason, then we should delete the new one. This is not done on reload(), so there is extra logic after that to handle this situation.

Reply:
Some fmt.print was added to see if the error expected by the test was indeed the one we,re receiving, which it is indeed.

So now I think that something during reload() is holding the new block open. I'm trying to debug this but it's quite complicated since it behaves differently on Linux 😅